projects/website/markdown/posts/2021-09-21-tuist-2.0/post.mdx

```diff
@@ -0,0 +1,148 @@
+tuist plugin test # Tests the plugin using the "swift" CLI
+tuist plugin archive # Creates an pre-compiled archive to install the plugin
+```
+
+Not only will third-party developers build for the platform, 
+but we'll also use it to extract some of the current commands that are more suitable to be opted into.
+We can't wait to see what developers will build upon this new plugins architecture.
+
+## Caching improvements
+
+Improving build times remains one of the most concerning areas when scaling up Xcode projects.
+Luckily, 
+our internal graph representation of projects positioned us to provide a solution without the complexity and maintenance of introducing a new build system like [Bazel](https://bazel.build/) entails.
+The idea is simple,
+the project is represented by a graph with targets as nodes,
+some of which can be replaced with their binary counterpart at generation time.
+The binaries are identified with a fingerprint that changes if the target or any of its dependencies changes.
+Sounds trivial, but it has a set of challenges.
+
+One of the challenges we are going through is getting a **deterministic and accurate fingerprint**.
+It's challenging because Xcode projects support bringing implicitness to the build process.
+For instance,
+a script build phase could affect the built artifact of the target without the Tuist's fingerprinting logic being aware of it.
+We can't detect implicitness,
+but we can provide APIs to make it explicit.
+Detecting sources of implicitness will require  close collaboration with developers.
+
+Another challenge is **generating a valid graph with binaries after the mutation.**
+The scenarios that we used as a reference are very ideal and,
+in most cases, distant from reality.
+There are many different flavors of Xcode projects out there that the mutation logic needs to handle gracefully.
+If our mutation logic doesn't support them, 
+developers might end up getting a project that doesn't compile,
+or even worse,
+they won't get a project at all.
+
+When we adventured ourselves into the caching land,
+we knew it was not going to be easy.
+Our goal towards 3.0 is to continue to work with developers to make caching **accessible and bulletproof**.
+
+## Third-party dependencies
+
+Before we introduced [`Dependencies.swift`](https://docs.tuist.io/guides/third-party-dependencies),
+adding third-party dependency dependencies to a Tuist project was non-standard,
+and in some scenarios,
+led to integration issues that surfaced at compilation time.
+Moreover,
+the integration of Swift Packages proposed by Apple proved to offer a poor developer experience.
+For example,
+the resolution of dependencies sometimes fails at launch time, 
+and it gets invalidated after deleting the derivd data directory.
+
+`Dependencies.swift` offered a standard solution across Carthage and Package dependencies
+and integrated dependencies into the projects' graphs to allow users to leverage existing Tuist like binary caching.
```

Comment:
Should we mention that this has the advantage to also allow to use tuist caching on dependencies?

Reply:
Yes. We should definitively mention that.